### PR TITLE
t-SNE code refactoring

### PR DESCRIPTION
**Pull request by Pahandrovich**

### CODE REVIEW

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp [hunk 1]
```diff
@@ -128,24 +129,24 @@ services::Status boundingBoxKernelImpl(DataType * posx, DataType * posy, const I
...
     else
-        box[3] = (box[3] * (1. - 1e-3));
+        box[3] = (box[3] * (1. - boxEps));
 
     //save results
     radius       = services::internal::max<cpu, DataType>(box[1] - box[0], box[3] - box[2]) * 0.5;
-    posx[nNodes] = (box[0] + box[1]) * 0.5;
-    posy[nNodes] = (box[2] + box[3]) * 0.5;
+    posX[nNodes] = (box[0] + box[1]) * 0.5;
+    posY[nNodes] = (box[2] + box[3]) * 0.5;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:143]: `0.5` refers to `double` precision data type. Please consider the following:
```suggestion
    posY[nNodes] = (box[2] + box[3]) * static_cast<DataType>(0.5);
```

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp [hunk 2]
```diff
@@ -160,17 +161,19 @@ services::Status qTreeBuildingKernelImpl(IdxType * child, const DataType * posx,
     DataType r     = 0.;
     DataType px    = 0.;
     DataType py    = 0.;
+    const DataType localEps = 1e-6;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:164]: What about using `std::numeric_limits<DataType>::epsilon()`. If you make it `constexpr ` that will guarantee that we do not depend on standard library.

```suggestion
    constexpr DataType localEps = std::numeric_limits<DataType>::epsilon();
```

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp [hunk 3]
```diff
@@ -512,10 +522,11 @@ services::Status repulsionKernelImpl(const DataType theta, const DataType eps, c
...
-    const IdxType sizeOfBlock = services::internal::min<cpu, IdxType>(256, N / nThreads + 1);
+    const IdxType sizeOfBlock = services::internal::min<cpu, IdxType>(blockOfRows, N / nThreads + 1);
     const IdxType nBlocks     = N / sizeOfBlock + !!(N % sizeOfBlock);
 
-    dq[0] = radiusdSquared / thetaSquared;
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(DataType, DataType(1), radiusSquared / thetaSquared);
+    dq[0] = radiusSquared / thetaSquared;
     for (auto i = 1; i < maxDepth; i++)
     {
         dq[i] = dq[i - 1] * 0.25;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:532]:
```suggestion
        dq[i] = dq[i - 1] * static_cast<DataType>(0.25);
```

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp [hunk 4]
```diff
@@ -691,24 +704,24 @@ services::Status integrationKernelImpl(const DataType eta, const DataType moment
...
         {
-            const DataType dx = exaggeration * attrx[i] - zNorm * repx[i];
-            const DataType dy = exaggeration * attry[i] - zNorm * repy[i];
+            const DataType dx = exaggeration * attrX[i] - zNorm * repX[i];
+            const DataType dy = exaggeration * attrY[i] - zNorm * repY[i];
             localSum[0] += dx * dx + dy * dy;
 
-            gx = (dx * (ux = oldForcex[i]) < DataType(0)) ? gainx[i] + 0.2 : gainx[i] * 0.8;
+            gx = (dx * (ux = oldForceX[i]) < DataType(0)) ? gainX[i] + 0.2 : gainX[i] * 0.8;
             if (gx < 0.01) gx = 0.01;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:712]: What a magical numbers?

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp (hunk 2 above)
Comment (Pahandrovich) [reply cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:165]: I didn't find similar uses of std in the old interfaces. Do you mind leaving it as it is?

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp (hunk 4 above)
Comment (Pahandrovich) [reply cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp]: changing `gain` aligned with scikit-learn and cuml.

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp [hunk 5]
```diff
@@ -365,21 +370,28 @@ services::Status summarizationKernelImpl(IdxType * count, IdxType * child, DataT
...
+                            m = mass[ch] + DataType(duplicates[ch]) - DataType(1);
                         }
                     }
                     else
                         m = (ch >= N) ? (cnt += count[ch], curMass[i]) : (cnt++, mass[ch]);
                     // add child's contribution
                     cm += m;
-                    px += posx[ch] * m;
-                    py += posy[ch] * m;
+                    //DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(DataType, posX[ch], m);
```
Comment (agorshk) [inline cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:380]: Are you sure that posX[ch] * m won't overflow

**Review by agorshk — APPROVED** [review]

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp (hunk 5 above)
Comment (Pahandrovich) [reply cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:380]: this is the setting for openmp parallelization. It was introduced in scikit-learn 1.1 and is always equal to 1 so far.

Comment (Pahandrovich) [reply cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:380]: One of the checks  broken python tests, I'm still investigate why

**Review by vlad-nazarov — APPROVED** [review]

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp (hunk 2 above)
Comment (KulikovNikita) [reply cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp:165]: Okay. But such magical constants can vary dramatically for different types. That is probably not enough for let say `float16` type

File: cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp (hunk 4 above)
Comment (KulikovNikita) [reply cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp]: What about using `constexpr` for such constant decalration?

**Pull request merged**